frontend/pkg/frontend/frontend.go

```diff
@@ -641,3 +644,50 @@ func (f *Frontend) ArmDeploymentPreflight(writer http.ResponseWriter, request *h
 
 	arm.WriteDeploymentPreflightResponse(writer, preflightErrors)
 }
+
+func (f *Frontend) logSubscriptionDifferences(oldSub, newSub *arm.Subscription) {
```

Comment:
Maybe we could pass a *slog.Logger directly as a parameter and do not define the method in the frontend struct as it has a lot of methods already and we do not use anything here rather than f.logger, so it could make sense to just pass what we need to simplify it a bit. We could also group the 
// Compare tenantId
and 
// Compare registeredFeatures code blocks
under the same parent "if" as it is repeated right now. Lastly, we could merge some split if in the same using &&.

Maybe something like this:
```go
func logSubscriptionDifferences(logger *slog.Logger, oldSub, newSub *arm.Subscription) {
	// Compare states
	if oldSub.State != newSub.State {
		logger.Info(fmt.Sprintf("Subscription state changed from %s to %s", oldSub.State, newSub.State))
	}

	if oldSub.Properties != nil && newSub.Properties != nil {
		// Compare tenantId
		if oldSub.Properties.TenantId != nil && newSub.Properties.TenantId != nil &&
			*oldSub.Properties.TenantId != *newSub.Properties.TenantId {
			logger.Info(fmt.Sprintf("Subscription tenantId changed from %s to %s", *oldSub.Properties.TenantId, *newSub.Properties.TenantId))
		}

		// Compare registeredFeatures
		if oldSub.Properties.RegisteredFeatures != nil && newSub.Properties.RegisteredFeatures != nil {
			oldFeatures := featuresMap(oldSub.Properties.RegisteredFeatures)
			newFeatures := featuresMap(newSub.Properties.RegisteredFeatures)

			for featureName, oldState := range oldFeatures {
				newState, exists := newFeatures[featureName]
				if !exists {
					logger.Info(fmt.Sprintf("Feature %s removed", featureName))
				} else if oldState != newState {
					logger.Info(fmt.Sprintf("Feature %s state changed from %s to %s", featureName, oldState, newState))
				}
			}
			for featureName, newState := range newFeatures {
				if _, exists := oldFeatures[featureName]; !exists {
					logger.Info(fmt.Sprintf("Feature %s added with state %s", featureName, newState))
				}
			}
		}
	}
}
```
Just an idea while reviewing it.

Reply:
Another possibility, since all the messages are Info-level, is return a slice of message strings and let the caller forward each string to `logger.Info`.  Then the function wouldn't need a `slog.Logger` argument at all, and also wouldn't necessarily need to live in `frontend.go`.